modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -295,139 +319,252 @@ public void finishFull(HybridTimestampTracker timestampTracker, UUID txId, boole
-            if (!commit) {
-                clo.apply(null).handle((ignored, err) -> {
+        // This is a finishing thread.
+        if (performingFinish.get()) {
+            Function<Void, CompletableFuture<Void>> finishAction = ignored ->
+                    makeFinishRequest(
+                            observableTimestampTracker,
+                            commitPartition,
+                            commit,
+                            enlistedGroups,
+                            txId,
+                            finishingStateMeta.txFinishFuture()
+                    );
+
+            runFinish(commit, tuple, finishAction);
+        }
+
+        // The method `runFinish` has a side effect on `finishInProgressFuture` future, it kicks off another future that will complete it.
+        return tuple.finishInProgressFuture;
+    }
+
+    private static void runFinish(boolean commit, TxContext tuple, Function<Void, CompletableFuture<Void>> finishAction) {
+        // Wait for commit acks first, then proceed with the finish request.
+        CompletableFuture<Void> finisher = commit ? tuple.waitNoInflights() : completedFuture(null);
+
+        finisher
+                .thenCompose(finishAction)
+                .handle((ignored, err) -> {
                     if (err == null) {
-                        tuple.finishFut.complete(null);
+                        tuple.finishInProgressFuture.complete(null);
                     } else {
-                        tuple.finishFut.completeExceptionally(err);
+                        tuple.finishInProgressFuture.completeExceptionally(err);
                     }
                     return null;
                 });
-            } else {
+    }
 
-                // All inflights have been completed before the finish.
-                if (tuple.inflights == 0) {
-                    tuple.waitRepFut.complete(null);
-                }
+    private CompletableFuture<Void> makeFinishRequest(
+            HybridTimestampTracker observableTimestampTracker,
+            TablePartitionId commitPartition,
+            boolean commit,
+            Map<TablePartitionId, Long> enlistedGroups,
+            UUID txId,
+            CompletableFuture<TransactionMeta> txFinishFuture
+    ) {
+        HybridTimestamp commitTimestamp = getCommitTimestamp(commit);
+        // In case of commit it's required to check whether current primaries are still the same that were enlisted and whether
+        // given primaries are not expired or, in other words, whether commitTimestamp is less or equal to the enlisted primaries
+        // expiration timestamps.
+        CompletableFuture<Void> verificationFuture =
+                commit ? verifyCommitTimestamp(enlistedGroups, commitTimestamp) : completedFuture(null);
+
+        return verificationFuture.handle(
+                        (unused, throwable) -> {
+                            boolean verifiedCommit = throwable == null && commit;
```

Comment:
Here we may get an axception close to that you declared above: TX_WAS_ABORTED_ERR.